### PR TITLE
"Introduction to ARC/ORC in Nim" guest post

### PR DESCRIPTION
**Pull request by Yardanico**
Initial version, please report any cases of inaccuracies/typos/bad phrasing.

Article name is up to change, maybe we can think of something better?

### CODE REVIEW

File: jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+---
+title: "What are ARC and ORC in Nim?"
+author: Danil Yarantsev (Yardanico)
+excerpt: "Nim is advancing towards more efficient
```
Comment (Yardanico) [inline jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md:4]: Not sure about this one, my original intro was "Hello everyone! In this post, I'll try to explain what are ARC and ORC and how they will affect performance or other aspects of Nim. I won't dive too much into writing Nim code with ARC/ORC in mind, but rather give a more-or-less high-level explanation." but I checked other guest posts and made this excerpt instead

File: jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+(contrary to most other tracing GCs which do global tracing).
+I don't have much to say about ORC, but it retains most of ARC's advantages
+except **determinism** (partially) - by default ORC has an adaptive threshold
+for collecting cycles, and **hard realtime** (partially), for the same reason.
+To enable ORC you would need to compile your program with ``--gc:orc``,
+but ORC will probably become Nim's default GC in the future.
+
+
+
+## I'm excited! How do I test them out?
```
Comment (Yardanico) [inline jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md:162]: This section will need to be changed if we plan to post the article after 1.4 is out

File: jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+Reference counting is one of the most popular algorithms for freeing unused
+resources of the program. **Reference count** for any managed (controlled by the runtime)
+reference is how many times that reference is used in other places.
+When that count becomes zero -> the reference and all of its underlying data is destroyed.
+
+The main difference of ARC compared to Nim GCs is that ARC is fully **deterministic** -
+the compiler automatically **injects** destructors when it deems that some variable
+(a string, sequence, reference, or something else) is no longer needed.
+In this sense, it's similar to C++ with its destructors (RAII).
+To show what I mean we can use Nim's ``expandArc`` introspection (will be available in Nim 1.4).
```
Comment (Yardanico) [inline jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md:52]: If we post the article after 1.4, need to change this

**Comment by Yardanico** [comment]
@Araq please review :)

**Comment by Yardanico** [comment]
Also see https://github.com/nim-lang/website/issues/231 for a styling bug

File: jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
+---
+title: "What are ARC and ORC in Nim?"
+author: Danil Yarantsev (Yardanico)
+excerpt: "Nim is advancing towards more efficient 
+memory management models. Let's talk about ARC/ORC 
+and see how they will change the way memory works in Nim."
```
Comment (Yardanico) [inline jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md:6]: Not sure about this one, my original intro was "Hello everyone! In this post, I'll try to explain what are ARC and ORC and how they will affect performance or other aspects of Nim. I won't dive too much into writing Nim code with ARC/ORC in mind, but rather give a more-or-less high-level explanation." but I checked other guest posts and made this excerpt instead

**Comment by Yardanico** [comment]
@ee7 can you please take a look as well? :)

**Review by ee7 — COMMENTED** [review]
Here's a quick review just for English correctness/style.

I wasn't concentrating so much on the actual content/structure/arguments, sorry. Nor that the Nim content was actually technically correct.

But one thing: maybe we should mention somewhere that the `--gc` option is badly named, and say that it's more like "memory management strategies" instead.

File: jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+In Nim cycle collection has been done by the mark & sweep phase of the `refc` GC,
+but it's better to use ARC as the foundation to make something better, so please welcome:
+
+
+## ORC - Nim's cycle collector
+ORC is an all-new Nim's cycle collector based on ARC.
+It can be considered a full-blown GC since it includes a local tracing phase
+(contrary to most other tracing GCs which do global tracing).
+I don't have much to say about ORC, but it retains most of ARC's advantages
```
[the remark this answers is not in the document]
Comment (Yardanico) [reply jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md:154]: Yeah I should add that :)

File: jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+At its core ARC is a memory management model based on Automatic Reference Counting
+(some people mistake Nim's ARC for Swift's ARC; however they have one big difference -
+ARC in Nim doesn't use atomic RC like in Swift) with destructors and move semantics.
+
+Reference counting is one of the most popular algorithms for freeing unused
+resources of the program. **Reference count** for any managed (controlled by the runtime)
+reference is how many times that reference is used in other places.
+When that count becomes zero -> the reference and all of its underlying data is destroyed.
+
+The main difference of ARC compared to Nim GCs is that ARC is fully **deterministic** -
```
[the remark this answers is not in the document]
Comment (Yardanico) [reply jekyll/_posts/2020-10-14-what-are-arc-and-orc-in-nim.md:48]: Did it a bit differently - wanted to separate ARC from GCs

**Comment by Yardanico** [comment]
@ee7 thanks a ton for all of the suggestions, I'm not a native English speaker so your suggestions also help me to improve my English :)

**Pull request merged**